File: src/syscall.c

inline long internal_syscall0(long nr) {
  unsigned long int resultvar;
  asm volatile (
      "syscall\n\t"
      : "=a" (resultvar)
      : "0" (nr)
      : "memory", "cc", "r11", "cx");
  return (long) resultvar;
}


inline long internal_syscall1(long nr, unsigned long a1) {
  unsigned long int resultvar;
  register unsigned long _a1 asm ("rdi") = a1;
  asm volatile (
      "syscall\n\t"
      : "=a" (resultvar)
      : "0" (nr), "r" (_a1)
      : "memory", "cc", "r11", "cx");
  return (long) resultvar;
}

inline long internal_syscall2(long nr, unsigned long a1, unsigned long a2) {
  unsigned long int resultvar;
    register unsigned long _a1 asm ("rdi") = a1;
    register unsigned long _a2 asm ("rsi") = a2;
    asm volatile (
        "syscall\n\t"
        : "=a" (resultvar)
        : "0" (nr), "r" (_a1), "r" (_a2)
        : "memory", "cc", "r11", "cx");
    return (long) resultvar;
}

inline long internal_syscall3(long nr, unsigned long a1, unsigned long a2, unsigned long a3) {
  unsigned long int resultvar;
    register unsigned long _a1 asm ("rdi") = a1;
    register unsigned long _a2 asm ("rsi") = a2;
    register unsigned long _a3 asm ("rdx") = a3;
    asm volatile (
        "syscall\n\t"
        : "=a" (resultvar)
        : "0" (nr), "r" (_a1), "r" (_a2), "r" (_a3)
        : "memory", "cc", "r11", "cx");
    return (long) resultvar;
}

inline long internal_syscall4(long nr, unsigned long a1, unsigned long a2, unsigned long a3, unsigned long a4) {
  unsigned long int resultvar;
    register unsigned long _a1 asm ("rdi") = a1;
    register unsigned long _a2 asm ("rsi") = a2;
    register unsigned long _a3 asm ("rdx") = a3;
    register unsigned long _a4 asm ("r10") = a4;
    asm volatile (
        "syscall\n\t"
        : "=a" (resultvar)
        : "0" (nr), "r" (_a1), "r" (_a2), "r" (_a3), "r"(_a4)
        : "memory", "cc", "r11", "cx");
    return (long) resultvar;
}

inline long internal_syscall5(long nr, unsigned long a1, unsigned long a2, unsigned long a3, unsigned long a4, unsigned long a5) {
  unsigned long int resultvar;
    register unsigned long _a1 asm ("rdi") = a1;
    register unsigned long _a2 asm ("rsi") = a2;
    register unsigned long _a3 asm ("rdx") = a3;
    register unsigned long _a4 asm ("r10") = a4;
    register unsigned long _a5 asm ("r8")  = a5;
    asm volatile (
        "syscall\n\t"
        : "=a" (resultvar)
        : "0" (nr), "r" (_a1), "r" (_a2), "r" (_a3), "r"(_a4), "r"(_a5)
        : "memory", "cc", "r11", "cx");
    return (long) resultvar;
}

inline long internal_syscall6(long nr, unsigned long a1, unsigned long a2, unsigned long a3, unsigned long a4, unsigned long a5, unsigned long a6) {
  unsigned long int resultvar;
    register unsigned long _a1 asm ("rdi") = a1;
    register unsigned long _a2 asm ("rsi") = a2;
    register unsigned long _a3 asm ("rdx") = a3;
    register unsigned long _a4 asm ("r10") = a4;
    register unsigned long _a5 asm ("r8")  = a5;
    register unsigned long _a6 asm ("r9")  = a6;
    asm volatile (
        "syscall\n\t"
        : "=a" (resultvar)
        : "0" (nr), "r" (_a1), "r" (_a2), "r" (_a3), "r"(_a4), "r"(_a5), "r"(_a6)
        : "memory", "cc", "r11", "cx");
    return (long) resultvar;
}

File: src/arch/x86_64/syscall.rs
// On x86-64, the following registers are used for args 1-6:
// arg1: %rdi
// arg2: %rsi
// arg3: %rdx
// arg4: %r10
// arg5: %r8
// arg6: %r9
//
// rax is used for both the syscall number and the syscall return value.
//
// rcx and r11 are always clobbered. syscalls can also modify memory. With the
// `asm!()` macro, it is assumed that memory is clobbered unless the nomem
// option is specified.
use core::arch::asm;

use super::syscalls::Sysno;

/// A syscall that takes 0 arguments.
#[inline]
pub unsafe fn syscall0(n: Sysno) -> usize {
    let mut ret: usize;
    asm!(
        "syscall",
        inlateout("rax") n as usize => ret,
        out("rcx") _, // rcx is used to store old rip
        out("r11") _, // r11 is used to store old rflags
        options(nostack, preserves_flags, readonly)
    );
    ret
}

/// A syscall that takes 1 argument.
#[inline]
pub unsafe fn syscall1(n: Sysno, arg1: usize) -> usize {
    let mut ret: usize;
    asm!(
        "syscall",
        inlateout("rax") n as usize => ret,
        in("rdi") arg1,
        out("rcx") _, // rcx is used to store old rip
        out("r11") _, // r11 is used to store old rflags
        options(nostack, preserves_flags)
    );
    ret
}

/// A syscall that takes 2 arguments.
#[inline]
pub unsafe fn syscall2(n: Sysno, arg1: usize, arg2: usize) -> usize {
    let mut ret: usize;
    asm!(
        "syscall",
        inlateout("rax") n as usize => ret,
        in("rdi") arg1,
        in("rsi") arg2,
        out("rcx") _, // rcx is used to store old rip
        out("r11") _, // r11 is used to store old rflags
        options(nostack, preserves_flags)
    );
    ret
}

/// A syscall that takes 3 arguments.
#[inline]
pub unsafe fn syscall3(
    n: Sysno,
    arg1: usize,
    arg2: usize,
    arg3: usize,
) -> usize {
    let mut ret: usize;
    asm!(
        "syscall",
        inlateout("rax") n as usize => ret,
        in("rdi") arg1,
        in("rsi") arg2,
        in("rdx") arg3,
        out("rcx") _, // rcx is used to store old rip
        out("r11") _, // r11 is used to store old rflags
        options(nostack, preserves_flags)
    );
    ret
}

/// A syscall that takes 4 arguments.
#[inline]
pub unsafe fn syscall4(
    n: Sysno,
    arg1: usize,
    arg2: usize,
    arg3: usize,
    arg4: usize,
) -> usize {
    let mut ret: usize;
    asm!(
        "syscall",
        inlateout("rax") n as usize => ret,
        in("rdi") arg1,
        in("rsi") arg2,
        in("rdx") arg3,
        in("r10") arg4,
        out("rcx") _, // rcx is used to store old rip
        out("r11") _, // r11 is used to store old rflags
        options(nostack, preserves_flags)
    );
    ret
}

/// A syscall that takes 5 arguments.
#[inline]
pub unsafe fn syscall5(
    n: Sysno,
    arg1: usize,
    arg2: usize,
    arg3: usize,
    arg4: usize,
    arg5: usize,
) -> usize {
    let mut ret: usize;
    asm!(
        "syscall",
        inlateout("rax") n as usize => ret,
        in("rdi") arg1,
        in("rsi") arg2,
        in("rdx") arg3,
        in("r10") arg4,
        in("r8")  arg5,
        out("rcx") _, // rcx is used to store old rip
        out("r11") _, // r11 is used to store old rflags
        options(nostack, preserves_flags)
    );
    ret
}

/// A syscall that takes 6 arguments.
#[inline]
pub unsafe fn syscall6(
    n: Sysno,
    arg1: usize,
    arg2: usize,
    arg3: usize,
    arg4: usize,
    arg5: usize,
    arg6: usize,
) -> usize {
    let mut ret: usize;
    asm!(
        "syscall",
        inlateout("rax") n as usize => ret,
        in("rdi") arg1,
        in("rsi") arg2,
        in("rdx") arg3,
        in("r10") arg4,
        in("r8")  arg5,
        in("r9")  arg6,
        out("rcx") _, // rcx is used to store old rip
        out("r11") _, // r11 is used to store old rflags
        options(nostack, preserves_flags)
    );
    ret
}
